static/compiler-picker.ts

```diff
@@ -0,0 +1,209 @@
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+import _ from 'underscore';
+import TomSelect from 'tom-select';
+
+import { ga } from './analytics';
+import * as local from './local';
+
+export class CompilerPicker {
+    static /*const*/ favoriteGroupName = '__favorites__';
+    static /*const*/ favoriteStoreKey = 'favCompilerIds';
+    static nextSelectorId = 1;
+    domRoot: JQuery;
+    domNode: HTMLElement;
+    eventHub: any /* typeof hub.createEventHub() */;
+    id: number;
+    compilerService: any;
+    onCompilerChange: (any) => any;
+    tomSelect: TomSelect | null;
+    lastLangId: string;
+    lastCompilerId: string;
+    compilerIsVisible: (any) => any; // TODO => bool probably
+    constructor(domRoot: JQuery, hub: any /* Hub */, langId: string, compilerId: string,
+        onCompilerChange: (any) => any, compilerIsVisible?: (any) => any) {
+        this.eventHub = hub.createEventHub();
+        this.id = CompilerPicker.nextSelectorId++;
+        this.domNode = domRoot.find('.compiler-picker')[0];
+        this.compilerService = hub.compilerService;
+        this.onCompilerChange = onCompilerChange;
+        this.eventHub.on('compilerFavoriteChange', this.onCompilerFavoriteChange, this);
+        this.tomSelect = null;
+        if (compilerIsVisible) {
+            this.compilerIsVisible = compilerIsVisible;
+        } else {
+            this.compilerIsVisible = () => true;
+        }
+
+        this.initialize(langId, compilerId);
+    }
+
+    close() {
+        // Quick note while I'm here: This function is never called. It probably should be. The conformance view
+        // bypasses this function and does compilerEntry.picker.tomSelect.close(); manually. This function is the
+        // only time this.tomSelect can be null, might be nice if we can get rid of that.
+        this.eventHub.unsubscribe();
+        if (this.tomSelect)
+            this.tomSelect.destroy();
+        this.tomSelect = null;
+    }
+
+    initialize(langId: string, compilerId: string) {
+        this.lastLangId = langId;
+        this.lastCompilerId = compilerId;
+
+        this.tomSelect = new TomSelect(this.domNode as any /*TODO*/, {
```

Comment:
How come it does not understand HTMLElement?
https://tom-select.js.org/docs/#glossary says that NodeDefinition accepts a HTMLElement

Reply:
Without the `any` cast I get the following error:

```ts
Argument of type 'HTMLElement' is not assignable to parameter of type 'string | TomInput'.
  Type 'HTMLElement' is missing the following properties from type 'TomInput': disabled, required, value, type, checkValidity
```

I am not sure of the cause of the discrepancy here

Reply:
Ok, just needed to look at it when I wasn't tired :P Needs to specifically be an HTMLSelectElement or HTMLInputElement